test/libraries/CallVerification.t.sol

```diff
@@ -0,0 +1,96 @@
+import "../../src/contracts/types/CallTypes.sol";
+
+contract CallVerificationTest is Test {
+    using CallVerification for UserMetaTx;
+    using CallVerification for BidData[];
+
+    function buildUserMetaTx() internal pure returns (UserMetaTx memory) {
+        return UserMetaTx({
+            from: address(0x1),
+            to: address(0x2),
+            deadline: 12,
+            gas: 34,
+            nonce: 56,
+            maxFeePerGas: 78,
+            value: 90,
+            control: address(0x3),
+            data: "data"
+        });
+    }
+
+    function builderSearcherMetaTx() internal pure returns (SearcherMetaTx memory) {
+        return SearcherMetaTx({
+            from: address(0x1),
+            to: address(0x2),
+            value: 12,
+            gas: 34,
+            nonce: 56,
+            maxFeePerGas: 78,
+            userCallHash: "userCallHash",
+            controlCodeHash: "controlCodeHash",
+            bidsHash: "bidsHash",
+            data: "data"
+        });
+    }
+
+    function buildBidData(uint256 n) internal pure returns (BidData[] memory) {
+        BidData[] memory bidData = new BidData[](n);
+        for (uint256 i = 0; i < n; i++) {
+            bidData[i] = BidData({token: address(0x1), bidAmount: i});
+        }
+        return bidData;
+    }
+
+    function testGetUserCallHash() public {
+        this._testGetUserCallHash(buildUserMetaTx());
+    }
+
+    function _testGetUserCallHash(UserMetaTx calldata userMetaTx) external {
+        assertEq(userMetaTx.getUserCallHash(), keccak256(abi.encode(userMetaTx)));
+    }
+
+    function testGetBidsHash() public {
+        // 1 bid
+        BidData[] memory bidData = buildBidData(1);
+        assertEq(bidData.getBidsHash(), keccak256(abi.encode(bidData)));
+
+        // multiple bids
+        bidData = buildBidData(3);
+        assertEq(bidData.getBidsHash(), keccak256(abi.encode(bidData)));
+    }
+
+    function testGetCallChainHash() public {
+        ProtocolCall memory protocolCall = ProtocolCall({to: address(0x1), callConfig: 1});
+        UserMetaTx memory userMetaTx = buildUserMetaTx();
+        SearcherCall[] memory searcherCalls = new SearcherCall[](2);
+        searcherCalls[0] = SearcherCall({
+            to: address(0x2),
+            metaTx: builderSearcherMetaTx(),
+            signature: "signature1",
+            bids: buildBidData(1)
+        });
+        searcherCalls[1] = SearcherCall({
+            to: address(0x3),
+            metaTx: builderSearcherMetaTx(),
+            signature: "signature2",
+            bids: buildBidData(2)
+        });
+        this._testGetCallChainHash(protocolCall, userMetaTx, searcherCalls);
+    }
+
+    function _testGetCallChainHash(
+        ProtocolCall calldata protocolCall,
+        UserMetaTx calldata userMetaTx,
+        SearcherCall[] calldata searcherCalls
+    ) external {
+        bytes32 expectedCallChainHash = 0x69c0833b2f37f7a0cb7d040aaa3f4654d841ebf21e6781b530a9c978d0c8cf09;
```

Comment:
So these tests all pass for me,

But I think we should try refactor it similarly to me comment in your [Factory tests PR](https://github.com/Polygon-Fast-Lane/atlas/pull/17#pullrequestreview-1628996873) - where we extract the logic needed to get to this hash, put it in an easy-to-use test utils function, so its more robust and readable if people are curious about the isolated process for producing this hash.

That said, maybe there is also value in having a hardcoded expected hash test as well as a test that aims to reproduce the hash from known inputs. So we could have both types of tests. What do you think?

Reply:
I don't exactly see the value in reproducing the computation in the test. The test would obviously succeed since we are comparing hashes that were built the exact same way.

Having a hardcoded value, ie. the exact result we want our function to produce make more sense to me. But like I said before, I understand that this might not be the best practice.

I committed your suggestion in https://github.com/Polygon-Fast-Lane/atlas/pull/16/commits/081a2eab906151b1d5a1615df3417408604dd7ba, for your review.

Reply:
Yeah, I see what you're saying. Feels slightly uncomfortable just comparing a hash against another hardcoded hash because if it's wrong there's no transparent indication of what part went wrong. But as you say, we don't really get much value by just copying the logic into another place and re-running the same process there.

I think this is just a special case where the nature of hashes makes it hard to transparently test and understand where in the process things broke if something ever changes.

Looks great as it is now though, good job!